src/include/ipc/dai.h

```diff
@@ -87,4 +89,18 @@ struct sof_ipc_dai_config {
 	};
 } __attribute__((packed));
 
+struct sof_ipc_multidai_config {
+	struct sof_ipc_cmd_hdr hdr;
+
+	/* physical protocol and clocking */
+	uint16_t format;	/**< SOF_DAI_FMT_ */
+	uint16_t reserved16;	/**< alignment */
+
+	uint32_t num_dais;
+	uint32_t reserved[5];
+
+	struct sof_ipc_stream_map stream_map;
+	struct sof_ipc_dai_config configs[0];
+} __attribute__((packed));
+
```

Comment:
sorry, I don't understand the logic or rationale. And GitHub makes it really hard to follow the discussion between @slawblauciak and @lgirdwood since the initial commits are gone.

From my perspective, if the MULTIDAI is listed in the dai_type, then its configuration should be passed to the firmware in the same way as other DAIs. I don't see why we would have a new IPC, this is extra work for no good reason.

Reply:
Yes, that was my initial proposal. If it's particularly inconvenient for you I can revert it.

Reply:
My understanding is that we would have a DAI, say SSP0 and be able to load multiple different HW configurations for the DAI in one IPC (with an index/ID for each config). This aligns to the ASoC topology ABI. It also allows loading all DAI configs in a single IPC.

Reply:
@lgirdwood I think we are talking about different problems.

a) The discussion here is to define one 'multi-DAI' config IPC, so that we can enable synchronized transport on e.g. SSP0 and SSP4, or ALH1.3 and ALH 2.5 (the integer part is the link ID and the decimal the PDI number). So we send one compound configuration, and the firmware takes care of distributing the values at the low-level (since the kernel only interacts with a multi-dai).

n) What you describe in your comment is the ability to send multiple configurations for the same DAI (which could be extended for the multi-dai). for this we would need in addition a 'config_id' and a means to identify which config_id to apply. For example this could be the case for an SSP with support for 16kHz in BT HFP mode and 48kHz in A2DP mode. The selection of the config could be based on the_params, but it's not clear to me if the selection if done in firmware or at the kernel level.

We *may* combine the two topics and define a new structure that works for
- a)
- b)
- all of the above.

It's probably worth talking tomorrow morning on this, we're probably hitting the limits of off-line design

Reply:
@plbossart ack, that's what I was thinking - the same structure should be usable for both (the encoded DAI ID in each struct can be used by FW to config both a) and b) even simultaneously with the same IPC)

Reply:
ok, but then the definitions are not quite right

1. there should be an explicit number of configs
2. the num_dais and stream_map are only relevant for the b) case, so don't they belong in the configs[] field?

I think we should change to have a multi-config DAI IPC, that's backwards compatible with the existing one (i.e. the kernel would use the multi-config defintions even for a single config), and we should hide the 'virtual' dai as a regular DAI.

Reply:
Sounds reasonable. If I understand correctly, we'd keep the dai config struct as it is, use one of the reserved fields to pass the number of configs and turn the union of dai-specific configs into a flex array.

Green light?

Reply:
Hm, actually. How exactly are we supposed to have it backwards compatible?
It would be easy if there was a free reserved field at the beginning of the structure, where we'd have the number of configs. But we don't.

The only way I see here is actually duplicating the generic config fields and putting them in a flex-array but I honestly don't like this at all...

Reply:
I guess so if I'm reading this correctly and it's backwards compatible.